src/operators/Join.cpp

```diff
@@ -0,0 +1,227 @@
+#include "Join.h"
+
+#include <utility>
+#include <arrow/compute/api.h>
+#include <arrow/compute/kernels/compare.h>
+#include <arrow/compute/kernels/match.h>
+#include <table/util.h>
+#include <iostream>
+#include <arrow/scalar.h>
+
+namespace hustle {
+namespace operators {
+
+Join::Join(std::shared_ptr<OperatorResult> prev,
+               JoinGraph graph) {
+
+    prev_result_ = std::move(prev);
+    graph_ = std::move(graph);
+
+}
+
+std::unordered_map<int64_t, int64_t> Join::build_hash_table
+(const std::shared_ptr<arrow::ChunkedArray>& col) {
+
+    arrow::Status status;
+    arrow::compute::FunctionContext function_context(
+            arrow::default_memory_pool());
+    std::unordered_map<int64_t, int64_t> hash(col->length());
+
+    int row_offset = 0;
+
+    for (int i=0; i<col->num_chunks(); i++) {
+        // TODO(nicholas): for now, we assume the join column is INT64 type.
+        auto chunk = std::static_pointer_cast<arrow::Int64Array>(
+                col->chunk(i));
+
+        for (int row=0; row<chunk->length(); row++) {
+            hash[chunk->Value(row)] = row_offset + row;
+        }
+        row_offset += chunk->length();
+    }
+
+    return hash;
+}
+
+std::vector<arrow::compute::Datum> Join::probe_hash_table
+(const std::shared_ptr<arrow::ChunkedArray>& probe_col) {
+
+    arrow::Status status;
+
+    arrow::Int64Builder new_left_indices_builder;
+    arrow::Int64Builder new_right_indices_builder;
+    std::shared_ptr<arrow::Int64Array> new_left_indices;
+    std::shared_ptr<arrow::Int64Array> new_right_indices;
+
+    // Probe phase
+    int row_offset = 0;
+    for (int i = 0; i < probe_col->num_chunks(); i++) {
+
+        // TODO(nicholas): for now, we assume the join column is INT64 type.
+        auto left_join_chunk = std::static_pointer_cast<arrow::Int64Array>(
+                probe_col->chunk(i));
+
+        for (int row = 0; row < left_join_chunk->length(); row++) {
```

Comment:
Will be really interesting to explore (again in a subsequent design refactor) if we can make a vectorized probe, sort of parallel to what I was trying to demonstrate [here](https://github.com/UWQuickstep/gift/blob/27ed9f38770818deb2707274a4474fcf996c6023/types/BaseType.cpp#L103). This would make this loop likely go much faster.